private/pkg/command/process.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package command
+
+import (
+	"context"
+	"os/exec"
+
+	"go.uber.org/multierr"
+)
+
+type process struct {
+	cmd          *exec.Cmd
+	doneCallback func()
+	wait         chan error
+	procErr      error
+}
+
+// newProcess monitors cmd and will call doneCallback when the process exits.
+func newProcess(cmd *exec.Cmd, doneCallback func()) *process {
+	p := &process{
+		cmd:          cmd,
+		doneCallback: doneCallback,
+		wait:         make(chan error),
+	}
+	go p.monitor()
+	return p
+}
+
+func (p *process) monitor() {
+	p.wait <- p.cmd.Wait()
+	p.doneCallback()
+}
+
+func (p *process) Wait(ctx context.Context) error {
+	select {
+	case err, ok := <-p.wait:
+		// Process exited
+		if ok {
+			p.procErr = err
+			close(p.wait)
```

Comment:
If a user double-calls `Wait`, the program will panic. It should be documented that `Wait` can be called exactly once, and if it is called twice, an error should result and not a panic: https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/os/exec/exec.go;l=882

Reply:
The select won't panic as the receiver is in the <value>, <success> := form. It's here so we can tell when a call is double called rather than block forever or panic.

I'll make it so it'll return a specific error rather than capturing the exit code. The capture seemed useful but a permanent error is fine with me.